src/math/Horner.cpp

```diff
@@ -0,0 +1,51 @@
+/* -*- c++ -*-
+ * Copyright (c) 2012-2018 by the GalSim developers team on GitHub
+ * https://github.com/GalSim-developers
+ *
+ * This file is part of GalSim: The modular galaxy image simulation toolkit.
+ * https://github.com/GalSim-developers/GalSim
+ *
+ * GalSim is free software: redistribution and use in source and binary forms,
+ * with or without modification, are permitted provided that the following
+ * conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions, and the disclaimer given in the accompanying LICENSE
+ *    file.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions, and the disclaimer given in the documentation
+ *    and/or other materials provided with the distribution.
+ */
+
+#include "math/Horner.h"
+
+namespace galsim {
+namespace math {
+
+    void Horner(double* x, const int nx, double* coef, const int nc, double* result)
+    {
+        // Start at highest power
+        double* c = coef + nc-1;
+        // Ignore any trailing zeros
+        while (*c == 0. && c > coef) --c;
+        // Repeatedly multiply by x and add next coefficient
+        for (int i=0; i<nx; ++i) result[i] = *c;
+        while(--c >= coef) {
+            for (int i=0; i<nx; ++i) result[i] = result[i]*x[i] + *c;
```

Comment:
Maybe another opportunity for SSE?  I think KNL also has fused multiply and add instructions if we wanted to try that.

Reply:
I'm pretty sure this is already in a format that the compiler will be able to turn this into good SSE commands for us, but I didn't try it to confirm, since just moving this calculation to C++ was already a huge improvement.

Reply:
Yeah, looks like you're right: https://godbolt.org/z/M1pvoV  I think the inner loop starts at 132.  I'm a bit surprised there's no FMA, but I'm sure this is still quite efficient.

Reply:
Update: I was wrong.  Putting in a custom SSE implementation for this did indeed shave off a few more seconds.  c0645a8ae

Reply:
That's on my laptop though.  Using mostly the standard python compiler options, which do not include `march`, but do include `-O3` and  `-msse2`.  So maybe it isn't being quite as good in that case as when it's compiling for knl.  I am kind of surprised that my inexpert SSE skillz are better than clang's.

Reply:
That's really interesting.  I just tried adding `-march=native` and `-mtune=native` on my laptop, and things actually got a little bit slower.

I also played a bit with changing the compiler in the compiler-explorer example.  GCC and icc both like to use the fused multiply and add; clang never does - even on trunk.  Everybody likes to use the 512-bit bit zmm registers though, which does make me wonder if we're leaving something on the table (for KNL, not on our laptops) by coding directly to the 128-bit SSE registers.  Maybe we should do some profiling on KNL directly?

Reply:
We might be leaving a little bit on the table, but `_horner2d` is down to only 4.2 seconds out of 96.6 seconds total for `PhaseScreenPSF._shoot`.  So only at most a couple percent further speed up could be possible.  We're probably at the point now where the SED stuff and the LSSTCam WCS calculations are now much taller poles than anything left in GalSim.

Reply:
Fair enough.